dev/PipsControl/PipsControl.cpp

```diff
@@ -0,0 +1,369 @@
+    options.AnimationDesired(true);
+    sender.StartBringIntoView(options);
+}
+
+
+void PipsControl::OnMaxDisplayedPagesChanged(const int oldValue) {
+    m_lastMaxDisplayedPages = oldValue;
+    auto const numberOfPages = NumberOfPages();
+    if (m_lastMaxDisplayedPages != numberOfPages) {
+        UpdateVerticalPips(numberOfPages, MaxDisplayedPages());
+    }
+}
+
+void PipsControl::OnNumberOfPagesChanged(const int oldValue)
+{
+    m_lastNumberOfPagesCount = oldValue;
+    const int numberOfPages = NumberOfPages();
+    if (SelectedPageIndex() > numberOfPages - 1 && numberOfPages > -1) {
+        SelectedPageIndex(numberOfPages - 1);
+    }
+    UpdateVerticalPips(numberOfPages, MaxDisplayedPages());
+    UpdateNavigationButtonVisualStates();
+}
+
+void PipsControl::OnSelectedPageIndexChange(const int oldValue)
+{
+    // If we don't have any pages, there is nothing we should do.
+    // Ensure that SelectedPageIndex will end up in the valid range of values
+    // Special case is NumberOfPages being 0, in that case, don't verify upperbound restrictions
+    if (SelectedPageIndex() > NumberOfPages() - 1 && NumberOfPages() > 0)
+    {
+        SelectedPageIndex(NumberOfPages() - 1);
+    }
+    else if (SelectedPageIndex() < 0)
+    {
+        SelectedPageIndex(0);
+    }
+    else {
+        // Now handle the value changes
+        m_lastSelectedPageIndex = oldValue;
+
+        // Fire value property change for UIA
+        if (const auto peer = winrt::FrameworkElementAutomationPeer::FromElement(*this).try_as<winrt::PipsControlAutomationPeer>())
+        {
+            winrt::get_self<PipsControlAutomationPeer>(peer)->RaiseSelectionChanged(m_lastSelectedPageIndex, SelectedPageIndex());
+        }
+        UpdateNavigationButtonVisualStates();
+        UpdateVerticalPips(NumberOfPages(), SelectedPageIndex());
+        RaiseSelectedIndexChanged();
+    }
+}
+
+void PipsControl::RaiseSelectedIndexChanged()
+{
+    const auto args = winrt::make_self<PipsControlSelectedIndexChangedEventArgs>(m_lastSelectedPageIndex, SelectedPageIndex());
+    m_selectedIndexChangedEventSource(*this, *args);
+}
+
+void PipsControl::MovePipIdentifierToElement(int index) {
+
+    if (NumberOfPages() != 0) {
+        if (const auto repeater = m_verticalPipsRepeater.get())
+        {
+            if (const auto element = repeater.TryGetElement(m_lastSelectedPageIndex).try_as<winrt::Button>()) {
+                element.Style(DefaultIndicatorStyle());
+            }
+            if (const auto element = repeater.GetOrCreateElement(index).try_as<winrt::Button>()) {
+                element.UpdateLayout();
+                element.Style(SelectedIndicatorStyle());
+                ScrollToCenterOfViewport(element);
+            }
+        }
+    }
+}
+
+void PipsControl::SetVerticalPipsSVMaxSize() {
+    auto pipHeight = unbox_value<double>(ResourceAccessor::ResourceLookup(*this, box_value(c_PipsControlButtonHeightPropertyName)));
```

Comment:
>c_PipsControlButtonHeightPropertyName [](start = 91, length = 37)

You shouldn't depend on this resource value, from code behind you have no idea if its actually being used or not. Instead you should just get the ActualHeight() of one of your pips. YOu might need to Measure() it depending on where in the cycle this is being called.